### PR TITLE
[Reviewer: Ellie] Fix Timer iterators

### PR DESCRIPTION
**Pull request by sebrexmetaswitch**
This fixes an issue whereby the TSIterator could miss some timers, which meant that not all timers were replicated to the correct nodes when a resync happened.

### Issue
Chronos stores timers in either the short wheel, the long wheel or the heap. Each wheel is divided into buckets, and all timers in a short wheel bucket pop at once. As time advances, the short wheel is advanced. Once the end of the wheel is reached, all timers from the current bucket in the long wheel are moved into the short wheel, the short wheel starts from the beginning again and the long wheel is advanced. Once the end of the long wheel is reached, timers are moved into the long wheel from the heap and the long wheel starts from the beginning again.

The ISIterator is used to iterate through all timers, in order, from a specified point in time. TSIterator has a short wheel iterator, a long wheel iterator and a heap iterator. Before this change, for each of the wheel iterators, we would only ever iterate from the current bucket to the end of the wheel, i.e. the iterators are treating the wheels as linear, not circular (they're assuming there's nothing of value "earlier" in the wheel than where we're pointing).

When a timer is added, the following decision is made:
- if it's due to pop within the period of the short wheel from now, then add it to the appropriate bucket in the short wheel
- else if it's due to pop within the period of the long wheel from now, then add it to the appropriate bucket in the long wheel
- else add it to the heap

This means that inserting timers is (correctly) treating the wheels as circular.

The result is that, if a timer is added to the short wheel, and then we request the iterator for timers after a time which falls into a later bucket of the short wheel, then the iterator will never find that timer.

### Fix
To fix this,  when iterating through the wheels (for an iterator starting at bucket `X`, with the current time pointing at bucket `T` and the wheel having `N` buckets), instead of the old behaviour of iterating from `X` to `N`, we now iterate from `X` to `(T-1)`.

However, this leaves another problem which is that a timer `T1` may have been added at a time that it went into the long wheel, then some time later a timer `T2` is added that is due to pop after `T1` but is added to the short wheel. At this point, `T1` is still in the long wheel because we haven't yet refilled the short wheel from the long (which we will do before `T1` pops, so that's fine). In this case, if we try to iterate through the timers, we'll get `T2` ***before*** `T1`, because items in the short wheel's iterator are returned first.

To fix this problem, we refill each of the wheels when creating the iterator (which is OK from a performance perspective because we only do this on re-sync). However, the normal refill method for the short wheel assumes that everything in the current bucket of the long wheel should be moved across. This isn't true in this case, because we're refilling the wheel from a point that may be halfway through the cycle, so there may be timers in the long wheel bucket that can't fit yet. In addition, we have to fill from the ***next*** bucket of the long wheel, rather than the current bucket, because we're actually looking ahead.

For this reason, we add a new method `refill_short_wheel_from_next_long_bucket` to the store, which does exactly this and is only used when creating the iterators.

Also, there were a couple of places where we were doing overflow-unsafe comparisons of timers, which we shouldn't do, so those are fixed up to.

### Testing done
 - added UTs to test these various edge cases, that failed before this fix and pass after
 - ensured `make full_test` passed
 - ensured the fv tests pass

### CODE REVIEW

**Comment by sebrexmetaswitch** [comment]
@eleanor-merry I've now merged the fixes to the FV tests into this branch and verified the FV tests pass on this branch, so this should be reviewable now

**Comment by sebrexmetaswitch** [comment]
One more commit that I figured I'd throw on here - we could occasionally crash because the _ordered_timers vector was cleared by the _iterator was not updated to point to _ordered_timers.end() afterwards

File: src/timer_store.cpp [hunk 1]
```diff
@@ -339,6 +340,31 @@ void TimerStore::refill_short_wheel()
   long_bucket->clear();
 }
 
+// Refill the short timer wheel by distributing timers from the next bucket in
+// the long timer wheel.
+// Because not all timers in the next bucket will fit into the short wheel, we
```
Comment (eleanor-merry) [inline src/timer_store.cpp]: Not sure about describing this as fitting in - can you explain more why some timers in the long bucket shouldn't go into the short bucket (e.g. only timers due to pop in SHORT_WHEEL_TIME should get moved).
